File: Data Structures & Algorithms/climbing-stairs/submission-4.py
class Solution:
    def climbStairs(self, n: int) -> int:
        dp = {}

        def backtracking(num):
            if(num > n):
                return 0
            if(num == n):
                return 1
            if(num in dp):
                return dp[num]

            
            dp[num] = backtracking(num + 2) + backtracking(num + 1)

            return dp[num]
        
        return backtracking(0)

File: Data Structures & Algorithms/top-k-elements-in-list/submission-4.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();

        for(int num: nums)
        {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }

        PriorityQueue<int[]> heap = new PriorityQueue<>((a,b) -> a[0] - b[0]);

        for(int num: map.keySet())
        {
            heap.add(new int[] {map.get(num), num});

            if(heap.size() > k)
            {
                heap.poll();
            }

        }

        int[] res = new int[k];

        for(int i = 0; i < k; i++)
        {
            res[i] = heap.poll()[1];
        }

        return res;

    }
}
